packages/SwingSet/test/reap-all/reap-all.test.js

```diff
@@ -43,41 +43,125 @@ test('reap all vats', async t => {
+    t.is(dump.acceptanceQueue.length, acceptanceLength);
+    t.is(dump.runQueue.length, runLength);
+    t.is(dump.reapQueue.length, expectedReapQueueSet.size);
+    const reapQueueActual = new Set(dump.reapQueue);
+    t.deepEqual(reapQueueActual, expectedReapQueueSet);
+  };
+
+  checkQueues([], 12);
 
   c.reapAllVats();
-  const dumpPreReap = c.dump();
-  t.is(dumpPreReap.acceptanceQueue.length, 12);
-  t.is(dumpPreReap.reapQueue.length, 11);
-  t.is(dumpBefore.runQueue.length, 0);
-  // prettier-ignore
-  const reapQueueReference =
-    new Set(['v1', 'v3', 'v6', 'v7', 'v8', 'v5', 'v2', 'v4', 'v9', 'v10', 'v11'])
-  const reapQueueActual = new Set(dumpPreReap.reapQueue);
-  t.deepEqual(reapQueueActual, reapQueueReference);
+  checkQueues(allVats, 12);
+
+  await c.run();
+  checkQueues([]);
+
+  // Create a chain of references
+
+  /** @type {string[]} */
+  const exportedPromises = [];
+  exportedPromises.push(
+    /** @type {string} */ (c.queueToVatRoot('bootstrap', 'getExport')),
+  );
+
+  for (let i = 0; i < 3; i += 1) {
+    const lastExported = exportedPromises.slice(-1)[0];
+    exportedPromises.push(
+      /** @type {string} */ (
+        c.queueToVatObject(dynamicRoots[i], 'makeHolder', [kslot(lastExported)])
+      ),
+    );
+  }
+  await c.run();
+
+  // Drop our interest in the intermediary promises without gaining an interest in their resolution
+  for (const p of exportedPromises.splice(0, exportedPromises.length - 1)) {
+    c.kpResolution(p, { incref: false });
+  }
+
+  // Workaround to trigger processRefcounts
+  c.queueToVatObject(
+    dynamicRoots.slice(-1)[0],
+    'doSomething',
+    ['ping'],
+    'none',
+  );
+
+  await c.run();
+  checkQueues([]);
 
+  const postMessagesReapPos = c.reapAllVats();
+  checkQueues(allVats);
+
+  await c.run();
+  checkQueues([]);
+
+  const postReapReapPos = c.reapAllVats(postMessagesReapPos);
+  // Verify that if there is nothing to do, vat positions don't change through reapAll
+  t.deepEqual(postReapReapPos, postMessagesReapPos);
+  checkQueues([]);
+
+  // Drop our interest in the last promise of the chain, dropping the refcount onto its resolution
+  c.kpResolution(exportedPromises.splice(0, 1)[0], { incref: false });
+  // Workaround to trigger processRefcounts, making sure we only add activity to the vat retiring an export
+  c.queueToVatObject(
+    dynamicRoots.slice(-1)[0],
+    'doSomething',
+    ['ping'],
+    'none',
+  );
   await c.run();
-  const dumpPostReap = c.dump();
-  t.is(dumpPostReap.acceptanceQueue.length, 0);
-  t.is(dumpPostReap.reapQueue.length, 0);
-  t.is(dumpBefore.runQueue.length, 0);
+  checkQueues([]);
+
+  // Continuously trigger reapAllVats for vats that have seen deliveries,
+  // asserting that for each round, the expected vat is reporting a reap.
+  // Because of our chained references, our GC shakes lose an object for
+  // each reap round.
+  let prevReapPos = postMessagesReapPos;
+  for (const vat of ['v11', 'v10', 'v9', 'v1']) {
```

Comment:
Accompanying the above suggestion:
```suggestion
  for (const vat of [dyn3, dyn2, dyn1, c.vatNameToID('bootstrap')]) {
```